File: Algorithms/Sorting_Algorithms/Circle_Sort/Circle_Sort.c
#include <stdio.h>
#include <stdlib.h>

int circle_sort_inner(int *start, int *end)
{
  int *p, *q, t, swapped;
 
  if (start == end) return 0;
  for (swapped = 0, p = start, q = end; p<q || (p==q && ++q); p++, q--)
    if (*p > *q)
      t = *p, *p = *q, *q = t, swapped = 1;
 
  return swapped | circle_sort_inner(start, q) | circle_sort_inner(p, end);
}

void circle_sort(int a[], int n) 
{ 
    while (circle_sort_inner(a, a + (n - 1))) 
    { 
       ; 
    } 
} 
 
int main(void)
{
  int N , i ,j;
  printf("Enter Size of Array :- \n");
  scanf("%d",&N);
  int A[N];
  for(i = 0 ; i < N;i++)
    scanf("%d",&A[i]);

  circle_sort(A, N);
  printf("Sorted Array :- \n");
  for(i = 0 ; i < N;i++)
    printf("%d ",A[i]);
  printf("\n");

  return 0;
}

File: Algorithms/Sorting_Algorithms/Circle_Sort/Circle_Sort.java
import java.util.Scanner;
class circleSortClass
{
  boolean circleSortInner(int a[], int low, int high) 
  {
      boolean swapped = false;
      if (low == high)
          return false;
      int l = low, h = high;
      int mid = (high - low) / 2;
      boolean left, right;
      while (l < h)
      {
	        if (a[l] > a[h])
	        {
	            swap (a,l,h);
	            swapped = true;
          }
	        l++;
	        h--;
      }
      if (l == h)
          if (a[l] > a[h + 1])
	        {
	            swap (a,l,h + 1);
  	        swapped = true;
	        }
          left = circleSortInner (a, low, low + mid);
          right = circleSortInner (a, low + mid + 1, high);
          return swapped || left || right;
  }
 

  void circleSort (int a[], int n) 
  {
      while (circleSortInner (a, 0, n - 1));
  }


  void swap(int arr[],int l,int h)
  {
      int temp=arr[l];
      arr[l]=arr[h];
      arr[h]=temp;
  }

 

  // Driver program 
  public static void main(String[] args)
  {
      int arr[];
      Scanner sc=new Scanner(System.in);
      System.out.println("Enter number of elements to be sorted:");
      int n=sc.nextInt();
      arr = new int[n]; 
      System.out.println("Enter "+n+" elements");
      for (int i=0;i<n;i++)
          arr[i]=sc.nextInt();
      circleSortClass c=new circleSortClass();
      c.circleSort (arr, n);
      System.out.println("Array after sorting:");
      for (int i = 0; i < n; i++)
          System.out.print(arr[i]+" ");
      System.out.println();
  }
}
